_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+* Handle errors that may occur during message consumption, determining when a message should be retried and when an error is considered non-recoverable. 
+
+**consumer configuration**
+
+We'll specify the minimal configuration properties required for a consumer instance:
+
+* `BOOTSTRAP_SERVERS_CONFIG` for connection to the Kafka brokers. This picks up the value of the `BOOTSTRAP_SERVERS` constant.
+* `CLIENT_ID_CONFIG` that uses a randomly generated UUID as a client ID for tracking the source of requests.
+* `ENABLE_AUTO_COMMIT_CONFIG` to disable automatic offset committing by Kafka. We will control offset commits in our application.
+* `AUTO_OFFSET_RESET_CONFIG` to define the starting point for consuming messages when no committed offset is found for the configured group ID and assigned partition.  
+* `KEY_DESERIALIZER_CLASS_CONFIG` and `VALUE_DESERIALIZER_CLASS_CONFIG` to specify deserializers that transform message keys and values into a suitable format. 
+In this case, we'll specify the `ByteArrayDeserializer` because it simply represents the keys and values as bytes without further transformation.
+
+> Using the earliest offset behavior avoids messages being lost when the offsets retention period (`offsets.retention.minutes`) configured for a broker has ended.
+
+Producers and consumers can use different serializers and deserializers as long as the consumer's deserializer is capable of correctly interpreting the serialized data produced by the producer's serializer.
+In our [producer example](https://strimzi.io/blog/2023/10/03/kafka-producer-client-essentials/), messages are generated as random byte arrays, which serve as the content of the messages sent to the Kafka cluster. Given this format, it makes sense to choose the `ByteArrayDeserializer` for deserialization.  
+
+We'll also include methods that help with these operations:
+
+**`sleep` method**
+
+- Introduces a delay in the message-receiving process for a specified number of milliseconds.
+- Useful for controlling the message consumption rate or simulating message processing times.
+- Handles potential `InterruptedException` if the thread responsible for sending messages is interrupted while paused.
+
+**`retriable` method**
+
+- Determines whether to retry message consumption and offset management following a message consuming error.
+- Returns `true` if the message consuming process can be retried. The Kafka consumer automatically handles retries for certain errors, such as connection errors.
+- Returns `false` for null and specified exceptions, or those that do not implement the `RetriableException` interface.
+- Customizable to include other errors and implementing retry logic for business level exceptions.
+
+> By default, Kafka operates with at-least-once message delivery semantics, which means that messages can be delivered more than once in certain scenarios, potentially leading to duplicates.
```

Comment:
In Kafka we typically avoid talking about "delivery semantics" as delivery sort of implies records flowing end to end from a producer to a consumer. Instead semantics are controlled on each side, and you can use different semantics. So it's more correct to just talk about `at-least-once semantics` which does not have that implied meaning.

Reply:
Same below, it's a complicated topic so when just scratching the surface it's safer to basically avoid using "delivery".

Reply:
removed both instances of `delivery semantics`